api.go

```diff
@@ -0,0 +1,169 @@
+
+	var t0, t1, t2, t3, t4 time.Time
+
+	trace := &httptrace.ClientTrace{
+		DNSStart: func(_ httptrace.DNSStartInfo) { t0 = time.Now() },
+		DNSDone:  func(_ httptrace.DNSDoneInfo) { t1 = time.Now() },
+		ConnectStart: func(_, _ string) {
+			if t1.IsZero() {
+				// connecting to IP
+				t1 = time.Now()
+			}
+		},
+		ConnectDone: func(net, addr string, err error) {
+			if err != nil {
+				log.Printf("unable to connect to host %v: %v", addr, err)
+			}
+			t2 = time.Now()
+		},
+		GotConn:              func(_ httptrace.GotConnInfo) { t3 = time.Now() },
+		GotFirstResponseByte: func() { t4 = time.Now() },
+	}
+
+	// We'll use our Context in this request in case we have to shut down midstream
+	req = req.WithContext(httptrace.WithClientTrace(ctx, trace))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Failed to read response:", err)
+		return
+	}
+
+	// Send our server response code as an event
+	storage.EventDistributor <- j.makeEvent(resp.StatusCode)
+
+	// Save response for use by other jobs
+	if _, ok := responses[j.Name]; ok {
+		responses[j.Name].Completed = true
+		responses[j.Name].Response = resp
+		log.Println("Response saved job:", j.Name)
+		go waitAndClearResponse(responses, j.Name)
+	}
+
+	// Even though we never read the response body, if we don't close it,
+	// the http.Transport goroutines will terminate and the app will eventually
+	// crash due to OOM
+	resp.Body.Close()
+
+	t5 := time.Now() // after read body
+	if t0.IsZero() {
+		// we skipped DNS
+		t0 = t1
+	}
+
+	url, err := url.Parse(j.URL)
+	if err != nil {
+		log.Println("Failed to parse URL:", err)
+		return
+	}
+
+	switch url.Scheme {
+	case "https":
+		storage.MetricDistributor <- j.makeMetric("dns_duration_milliseconds", t1.Sub(t0).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_connection_duration_milliseconds", t2.Sub(t1).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("tls_handshake_duration_milliseconds", t3.Sub(t2).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_processing_duration_milliseconds", t4.Sub(t3).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_response_duration_milliseconds", t5.Sub(t4).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("time_to_first_byte_milliseconds", t4.Sub(t0).Seconds()*1000)
+
+	case "http":
+		storage.MetricDistributor <- j.makeMetric("dns_duration_milliseconds", t1.Sub(t0).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_connection_duration_milliseconds", t3.Sub(t1).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_processing_duration_milliseconds", t4.Sub(t3).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_response_duration_milliseconds", t5.Sub(t4).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("time_to_first_byte_milliseconds", t4.Sub(t0).Seconds()*1000)
+	}
+}
+
+func waitAndClearResponse(responses map[string]*JobResponse, name string) {
```

Comment:
Completed indicates whether the job has completed. Think of it as "valid" when dealing with nullible values. Also, just an FYI, this isn't super scalable and depending on latency, this could very well break, so we will need to explore options on how to change this. I have 2 in mind that I want to talk to you about.